File: javaalgo/src/kakao3.java
import java.util.HashSet;
import java.util.Set;

public class kakao3 {
	static String[][] input = {{"100","ryan","music","2"},{"200","apeach","math","2"},{"300","tube","computer","3"},{"400","con","computer","4"},{"500","muzi","music","3"},{"600","apeach","music","2"}};
	public static void main(String[] args) {
		solution(input);
	}
	
	static int col, row;
	
	
	static int solution(String[][] relation) {
        
		
        int total = 1<<relation[0].length;
        Set<Integer> keys = new HashSet<>();
        for(int i=1; i<total; ++i) {
        	boolean isMin = true;
        	for(int uniq: keys) {
        		if((uniq & i) == uniq) {
        			isMin = false;
        			break;
        		}
        	}
        	if(!isMin) continue;
        	int t = i;
        	int col = 0;
        	StringBuilder[] strs = new StringBuilder[relation.length];
        	for(int k=0; k<strs.length; ++k) {
        		strs[k] = new StringBuilder();
        	}
        	while(t>0) {
        		if((t&1) == 1) {
        			for(int k=0; k<relation.length; ++k) {
        				strs[k].append(relation[k][col]);
        			}
        		}
        		t >>= 1;
        		col++;
        	}
        	Set<String> set = new HashSet<>();
        	for(StringBuilder sb: strs) {
        		set.add(sb.toString());
        	}
        	if(set.size() == relation.length) {
        		keys.add(i);
        	}
        	
        	
        }
        System.out.println(keys.size());
        return keys.size();
    }
	
}

File: 0819/kakao3.py


def solution(relation):
    row = len(relation)
    col = len(relation[0])
    key = [[]for i in range(col)]

    
    for i in range(1, col+1):
        makecols(i, set(), 0, 0, col, key, row, relation)

    print(key)


def check(row, cols, relation, key, now):
    tmp = set()
    for r in range(row):
        T = ""
        for c in cols:
            T += str(relation[r][c])
        tmp.add(T)
        print(T)
    if len(tmp)==row:
        key[now-1].append(cols)

def makecols(MAX, ret, cnt, now, col, key, row, relation):
    if ret in key[len(ret)-2]:
        return
    if MAX == cnt:
        check(row, ret, relation, key, MAX)
        return
    if cnt+now >= col:
        return
    for i in range(now, col):
        ret.add(i)
        makecols(MAX, ret, cnt+1, i+1, col, key, row, relation)


input = [["100","ryan","music","2"],["200","apeach","math","2"],["300","tube","computer","3"],["400","con","computer","4"],["500","muzi","music","3"],["600","apeach","music","2"]]

solution(input)